lib/Parse/Lexer.cpp

```diff
@@ -1316,6 +1340,62 @@ static const char *skipToEndOfInterpolatedExpression(const char *CurPtr,
   }
 }
 
+/// extract content of string literal from inside quotes
+static StringRef getStringLiteralContent(const Token &Str) {
+  StringRef Bytes = Str.getText();
+  Bytes = Bytes.drop_front().drop_back();
+
+  if (Str.getStringModifiers() & StringLiteralMultiline) {
+    Bytes = Bytes.drop_front().drop_back();
+    Bytes = Bytes.drop_front().drop_back();
+  }
+
+  return Bytes;
+}
+
+/// determine contents of literal to be normalised - either
+/// to strip indenting or normalise line endings to a single \n
+static std::string getTrailingIndentOrWindowsLineEnding(const Token &Str) {
+  StringRef Bytes = getStringLiteralContent(Str);
+  const char *end = Bytes.end(), *start = end;
+
+  // work back from the end to find whitespace to strip
+  while (start > Bytes.begin() && isWhitespace(start[-1])) {
+    if (*--start == '\n' || *start == '\r') {
+      if (start[-1] == '\r')
```

Comment:
This is not right, I think: \r\r is two classic Mac newlines, not one. Surely, what you want is this:

```C++
switch (*--start) {
case '\n':
  if (start[-1] == '\r')
    --start;
  LLVM_FALLTHROUGH;
case '\r':
  return std::string(start, end-start);
}
```